### PR TITLE
add basic functionality and styles for popover

### LINKED ISSUE #2: Create an Accessible Github Page for Demo Purposes
Create an accessible github page that shows a demo for the popover.

### PR DESCRIPTION
**Pull request by craigmichaelmartin**
@akersinformz

### High level:
Popover ng2 directive that can be used like this
```html
<span
  suprePopoverHeader="How is that possible?"
  suprePopoverBody="The Browns have a history of losing to the bye week"
  suprePopoverPosition="right">
    The Browns are likely to go 0-17 this year.
</span>
```

### Implementation details
- Draws extensively from ng2-popover. I would love to use it in a way that keeps
ng2-popover as an upstream lib (especially for the code it reconstructs from Tether for positioning the popover) - where code can be shared/test/perfected together.
- Uses bootstrap's popover scss. This allows us to inherit the wealth of bug fixes, tests, etc from upstream (as ideally should be the case for ng2-popover).

### Installation
```bash
npm i -S CINBCUniversal/suprematism-popover
```
Until it is published to npm, point to github. A consequence of this is that built files must be checked-in. When we start publishing to npm with `npm publish`, there is a prehook to build the files and a posthook to delete them (so that only source files are saved in git). For now, after doing development, we must manually run the publish prehook and check-in the files.

This PR closes #2

### CODE REVIEW

**Comment by craigmichaelmartin** [comment]
Writing more tests, found an issue with the popover affecting the flow of elements after; resolving now.

File: example/app/app.component.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { Component } from '@angular/core';
+
+@Component({
+  selector: 'supre-root',
```
Comment (akersinformz) [inline example/app/app.component.ts]: For readability lets move the template into a html file

File: package.json [hunk 2]
```diff
@@ -39,6 +39,13 @@
...
     "tslint": "3.13.0",
-    "typescript": "2.0.2"
+    "typescript": "2.0.2",
+    "zone.js": "^0.6.23"
+  },
+  "peerDependencies": {
+    "@angular/core": "^2.0.0"
+  },
+  "devDependencies": {
+    "bootstrap": "^4.0.0-alpha.4"
```
Comment (akersinformz) [inline package.json]: What is dependent on bootstrap?

File: src/popover.component.ts [hunk 3]
```diff
@@ -0,0 +1,240 @@
+import { Component, Input, AfterViewInit, ElementRef, ChangeDetectorRef,
+         ViewChild} from '@angular/core';
+import { PopoverDirective } from './popover.directive';
+import { PopoverPositionType } from './popover-position.type';
+
+@Component({
+  selector: 'supre-popover',
+  template: `
```
Comment (akersinformz) [inline src/popover.component.ts]: For readability lets move the template into a html file

**Comment by craigmichaelmartin** [comment]
Ready.

File: e2e/app.po.ts [hunk 4]
```diff
@@ -1,11 +1,66 @@
 import { browser, element, by } from 'protractor';
 
-export class SuprematismTooltipPage {
+export class SuprematismPopoverPage {
+
+  hostMap = new Map([
+    ['top', {getElement: () => this.$('.js-topElement')}],
+    ['right', {getElement: () => this.$('.js-rightElement')}],
```
Comment (akersinformz) [inline e2e/app.po.ts]: ES6 Map - Nice!

File: package.json (hunk 2 above)
Comment (craigmichaelmartin) [inline package.json]: The component extends bootstrap's popover scss. This allows us to inherit the wealth of bug fixes, tests, browser compat, etc from upstream.

Comment (akersinformz) [inline package.json]: Ok.  We might what to decouple ourselves from bootstrap in the future.  I suggest you fix the version to "bootstrap": "4.0.0-alpha.4" so we don't run into issues with a new version breaking the popover.

Comment (craigmichaelmartin) [inline package.json]: Bootstrap is abstracted away via the directives. Whatever is using the popover has no knowledge of the implementation. Agreed on locking it down.

**Comment by craigmichaelmartin** [comment]
Thanks for your feedback! PR updated.

File: example/app/app.module.ts [hunk 5]
```diff
@@ -1,18 +1,17 @@
 import { BrowserModule } from '@angular/platform-browser';
 import { NgModule } from '@angular/core';
-import { FormsModule } from '@angular/forms';
-import { HttpModule } from '@angular/http';
 
 import { AppComponent } from './app.component';
+// import { PopoverDirective } from '../../src/popover.directive';
```
Comment (caffeinelogic) [inline example/app/app.module.ts]: If this is not needed, let remove it

File: out-tsc/popover.component.d.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import { AfterViewInit, ElementRef, ChangeDetectorRef } from '@angular/core';
+import { PopoverDirective } from './popover.directive';
+import { PopoverPositionType } from './popover-position.type';
+export declare class PopoverComponent implements AfterViewInit {
+    private element;
+    private cdr;
+    popover: PopoverDirective;
```
Comment (caffeinelogic) [inline out-tsc/popover.component.d.ts]: Style question:
As a team do we want to put the public modifiers on these? I know its implied, but I think John Pap advocates for doing this for consistency etc. 

Thoughts? 

@craigmichaelmartin @akersinformz

File: src/index.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+export * from './popover.component';
+export * from './popover-position.type';
+
+@NgModule({
+    imports: [
+        CommonModule
+    ],
+    declarations: [
+        PopoverComponent,
+        PopoverDirective,
```
Comment (caffeinelogic) [inline src/index.ts]: Do we care about trailing commas? Should we create a tslint/eslint file somewhere we can all share?

Comment (akersinformz) [inline src/index.ts]: Style preference in ES6 world.  I am good either way.  Trailing commas is a ES6 pattern.

File: src/popover.component.ts [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+      ['border' + this.getPositionTitleCase() + 'Color']: this.color
+    };
+  }
+
+
+  // Private Methods
+  // I really wish all these next methods were references to the upstream 
+  // methods in ng2-popover, but cannot figure out how to make it so through
+  // inheritance or composition, given their being private.
+  // I'm still trying to figure out a way.
```
Comment (caffeinelogic) [inline src/popover.component.ts]: Do we want to have a chat about this, can we help you talk through it?

File: out-tsc/popover.component.d.ts (hunk 6 above)
Comment (craigmichaelmartin) [inline out-tsc/popover.component.d.ts]: I'm happy to remove `public` and let it be implied (or else add it the places its missing), given that 1) in general we want to enforce preference stuff like this, 2) we all agree on the preference for this case, and 3) we automate the check with eslint/tslint resulting in a very tight (ms) feedback loop through a linter, vs the feedback loop being through PRs, at that point it is much too late. I don't have a preference for 1,2. Happy with whatever the team wants, as long as 3.

File: src/popover.component.ts [hunk 9]
```diff
@@ -0,0 +1,216 @@
+import { Component, Input, AfterViewInit, ElementRef, ChangeDetectorRef,
+         ViewChild} from '@angular/core';
+import { PopoverDirective } from './popover.directive';
+import { PopoverPositionType } from './popover-position.type';
+
+@Component({
+  selector: 'supre-popover',
+  template: require('./popover.component.html'),
```
Comment (akersinformz) [inline src/popover.component.ts]: Here is the syntax for using a relative path to an external file without using required

templateUrl: 'popover.component.html',
styleUrls: ['popover.component.css']

Comment (craigmichaelmartin) [inline src/popover.component.ts]: I tried that first but couldn't get it to work, and I'm not sure why. I've got a long list of items I want to un-blackbox once I start getting a better understanding of ng and do right/better. This is one of them. Do you know what could be messing up the path? Depending on different things I've tried, I either get 404's in the browser, or 200's returning index.html.

File: src/popover.component.ts [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    const boundingClientRect = nativeEl.getBoundingClientRect();
+    return {
+      width: boundingClientRect.width || nativeEl.offsetWidth,
+      height: boundingClientRect.height || nativeEl.offsetHeight,
+      top: elBCR.top - offsetParentBCR.top,
+      left: elBCR.left - offsetParentBCR.left
+    };
+  }
+
+  private offset(nativeEl: any): { width: number, height: number, top: number, left: number } {
```
Comment (akersinformz) [inline src/popover.component.ts]: Missing type HTMLElement
i.e. offset(nativeEl: HTMLElement)

Comment (akersinformz) [inline src/popover.component.ts]: For readability you might what to setup a return interface/type for 
{ width: number, height: number, top: number, left: number }
Looks like it is used for placement and offset

File: src/popover.component.ts [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+    return {
+      width: boundingClientRect.width || nativeEl.offsetWidth,
+      height: boundingClientRect.height || nativeEl.offsetHeight,
+      top: boundingClientRect.top + (window.pageYOffset || window.document.documentElement.scrollTop),
+      left: boundingClientRect.left + (window.pageXOffset || window.document.documentElement.scrollLeft)
+    };
+  }
+
+  private getStyle(nativeEl: HTMLElement, cssProp: string): string {
+    if ((nativeEl as any).currentStyle) { // IE
```
Comment (akersinformz) [inline src/popover.component.ts]: missing HTMLELement type

File: src/popover.component.ts [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+    // finally try and get inline style
+    return (nativeEl.style as any)[cssProp];
+  }
+
+  private isStaticPositioned(nativeEl: HTMLElement): boolean {
+    return (this.getStyle(nativeEl, 'position') || 'static' ) === 'static';
+  }
+
+  private parentOffsetEl(nativeEl: HTMLElement): any {
+    let offsetParent: any = nativeEl.offsetParent || window.document;
```
Comment (akersinformz) [inline src/popover.component.ts]: I suggest using an OR for the type here HTMLELement | Document

File: src/popover.component.ts [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+          top: hostElPos.top - targetElHeight,
+          left: shiftWidth[pos1]()
+        };
+        break;
+    }
+
+    return targetElPos;
+  }
+
+  private placement(nativeEl: HTMLElement): { width: number, height: number, top: number, left: number } {
```
Comment (akersinformz) [inline src/popover.component.ts]: I am not sure I understand the parameter sent to this method.  I am probably just missing something.

Can this method receive an element or document?

I see you are testing against window.document

Out of the IF block you are accessing methods that would not exist on window.document if the element is a document object.  i.e. getBoundingClientRect or offsetWidth

File: src/popover.component.ts (hunk 9 above)
Comment (craigmichaelmartin) [inline src/popover.component.ts]: Sorry, I should have been more specific. The issue is when it is required in another project. 

screen shot 2016-10-18 at 10 21 28 am

The main property in package.json points to `out-tsc/index.js`, and `popover.component.html` and `popover.component.css` are right next to it.

screen shot 2016-10-18 at 10 23 33 am

**Comment by craigmichaelmartin** [comment]
Code is deployed to github pages: https://cinbcuniversal.github.io/suprematism-popover/
The PR closes #2

**Comment by craigmichaelmartin** [comment]
@akersinformz @caffeinelogic After submitting a PR to make private members protected, this component and directive now extend (ng2-popover)[https://github.com/pleerock/ng2-popover] and act as a facade, creating a sturdy directive interface we control, while opting to delegate much of the heavy lifting positioning implementation back to ng2-popover. We can swap out this implementation as desired.

I also removed `public` from modifiers (I've have setting up a base tslint config on my list of nice-to-get-to's). Let me know if there is anything else you would like to see for this PR.

**Pull request merged**